sites/game-client/map-renderer.mjs

```diff
@@ -211,6 +226,167 @@ class MapRenderer {
+    if (this.outOfBoundCanvas.width === newWidth && this.outOfBoundCanvas.height === newHeight) {
+      return;
+    }
+
+    // update canvas size
+    // note that the size should be larger to provide the margin of translation
+    this.outOfBoundCanvas.width = newWidth;
+    this.outOfBoundCanvas.height = newHeight;
+
+    // draw
+    const outerSpaceRenderInfo = this.map.graphicAsset.getTile(...OUTER_SPACE_TILE);
+    for (let y = 0; y < this.outOfBoundCanvas.height; y += MAP_CELL_SIZE) {
+      for (let x = 0; x < this.outOfBoundCanvas.width; x += MAP_CELL_SIZE) {
+        this.outOfBoundCtx.drawImage(
+            outerSpaceRenderInfo.image,
+            outerSpaceRenderInfo.srcX,
+            outerSpaceRenderInfo.srcY,
+            outerSpaceRenderInfo.srcWidth,
+            outerSpaceRenderInfo.srcHeight,
+            x,
+            y,
+            MAP_CELL_SIZE,
+            MAP_CELL_SIZE,
+        );
+      }
+    }
+  }
+
+  /**
+   * Render the background map to background canvas.
+   * This function is not reentrant (guarded by the if-else at the entry of the function).
+   */
+  async generateBackground() {
+    // There should be at most one thread doing this job in client's browser.
+    if (this._generateBackgroundLock !== undefined) {
+      return;
+    }
+    this._generateBackgroundLock = true;
+
+    // If the current map is not changed, no need to rerender.
+    const mapName = this.viewerPosition.mapName;
+    if (this._currentBackgroundMapName === mapName) {
+      this._generateBackgroundLock = undefined;
+      return;
+    }
+
+    // set canvas size
+    const mapSize = this.map.getMapSize(mapName);
+    this.backgroundCanvas.width = mapSize.width * MAP_CELL_SIZE;
+    this.backgroundCanvas.height = mapSize.height * MAP_CELL_SIZE;
+
+    // draw background
+    for (const [, layerName, renderFunction, renderArgs] of this.customizedBackgroundLayers) {
+      if (renderFunction === '_drawWatermark') {
+        this._drawWatermark(renderArgs);
+      } else {
+        // draw layer
+        for (let mapY = 0; mapY < mapSize.height; ++mapY) {
+          for (let mapX = 0; mapX < mapSize.width; ++mapX) {
+            const renderInfo = this.map.getCellRenderInfo(new MapCoord(mapName, mapX, mapY), layerName);
+            if (renderInfo === null) continue;
+
+            const canvasX = mapX * MAP_CELL_SIZE;
+            const canvasY = this.backgroundCanvas.height - renderInfo.srcHeight - mapY * MAP_CELL_SIZE;
+            this.backgroundCtx.drawImage(
+                renderInfo.image,
+                renderInfo.srcX,
+                renderInfo.srcY,
+                renderInfo.srcWidth,
+                renderInfo.srcHeight,
+                canvasX,
+                canvasY,
+                MAP_CELL_SIZE,
+                MAP_CELL_SIZE,
+            );
+          }
+
+          // avoid blocking the event loop
+          await new Promise((resolve) => {
+            setTimeout(resolve, 0);
```

Comment:
Optional: I would recommend doing double stencil so that it won't be possible for the user to see partially drawn background.